lib/internal/fs/copy/copy.js

```diff
@@ -0,0 +1,468 @@
+    futimes(fd, atime, mtime, (futimesErr) => {
+      close(fd, (closeErr) => {
+        if (callback) callback(futimesErr || closeErr);
+      });
+    });
+  });
+}
+
+function onDir(srcStat, destStat, src, dest, opts, cb) {
+  if (!destStat) return mkDirAndCopy(srcStat.mode, src, dest, opts, cb);
+  return copyDir(src, dest, opts, cb);
+}
+
+function mkDirAndCopy(srcMode, src, dest, opts, cb) {
+  mkdir(dest, (err) => {
+    if (err) return cb(err);
+    copyDir(src, dest, opts, (err) => {
+      if (err) return cb(err);
+      return setDestMode(dest, srcMode, cb);
+    });
+  });
+}
+
+function copyDir(src, dest, opts, cb) {
+  readdir(src, (err, items) => {
+    if (err) return cb(err);
+    return copyDirItems(items, src, dest, opts, cb);
+  });
+}
+
+function copyDirItems(items, src, dest, opts, cb) {
+  const item = items.pop();
+  if (!item) return cb(null);
+  return copyDirItem(items, item, src, dest, opts, cb);
+}
+
+function copyDirItem(items, item, src, dest, opts, cb) {
+  const srcItem = join(src, item);
+  const destItem = join(dest, item);
+  checkPaths(srcItem, destItem, opts, (err, stats) => {
+    if (err) return cb(err);
+    const { destStat } = stats;
+    startCopy(destStat, srcItem, destItem, opts, (err) => {
+      if (err) return cb(err);
+      return copyDirItems(items, src, dest, opts, cb);
+    });
+  });
+}
+
+function onLink(destStat, src, dest, opts, cb) {
+  readlink(src, (err, resolvedSrc) => {
+    if (err) return cb(err);
+    // TODO(bcoe): I don't know how this could be called, because
+    // getStatsForCopy will have used stat. Ask during review.
+    if (opts.dereference) {
+      resolvedSrc = resolve(process.cwd(), resolvedSrc);
+    }
+
+    if (!destStat) {
+      return symlink(resolvedSrc, dest, cb);
+    }
+    readlink(dest, (err, resolvedDest) => {
+      if (err) {
+        // Dest exists and is a regular file or directory,
+        // Windows may throw UNKNOWN error. If dest already exists,
+        // fs throws error anyway, so no need to guard against it here.
+        if (err.code === 'EINVAL' || err.code === 'UNKNOWN') {
+          return symlink(resolvedSrc, dest, cb);
+        }
+        return cb(err);
+      }
+      if (opts.dereference) {
+        resolvedDest = resolve(process.cwd(), resolvedDest);
+      }
+      if (isSrcSubdir(resolvedSrc, resolvedDest)) {
+        return cb(new ERR_FS_COPY_TO_SUBDIRECTORY({
+          code: 'EINVAL',
+          message: `cannot copy ${resolvedSrc} to a subdirectory of self ` +
+            `${resolvedDest}`,
+          path: dest,
+          syscall: 'copy',
+          errno: EINVAL,
+        }));
+      }
+
+      // Do not copy if src is a subdir of dest since unlinking
+      // dest in this case would result in removing src contents
+      // and therefore a broken symlink would be created.
+      // TODO(bcoe): I'm having trouble exercising this code in test,
+      // ask about during review.
+      if (destStat.isDirectory() &&
```

Comment:
@manidlou @RyanZim, I'm having a heck of a time figuring out a test case that exercises this code.

Reply:
I _think_ this is the test case:

https://github.com/jprichardson/node-fs-extra/blob/e6a95058c930953113177c9518f57e83cace3e79/lib/copy/__tests__/copy-prevent-copying-into-itself.test.js#L282-L296

Reply:
This turned out to be a bug I believe, the logic works correctly in `copySync` because of it uses `stat`:

```js
  if (statSync(dest).isDirectory() && isSrcSubdir(resolvedDest, resolvedSrc)) {
    throw new ERR_FS_COPY_SYMLINK_TO_SUBDIRECTORY({
      code: 'EINVAL',
      message: `cannot overwrite ${resolvedDest} with ${resolvedSrc}`,
      path: dest,
      syscall: 'copy',
      errno: EINVAL,
    });
  }
```

But in the asynchronous version `destStat` will have used `lstat` to dereference the symlink, so can never be a directory. I fixed this bug:

```js
      // Do not copy if src is a subdir of dest since unlinking
      // dest in this case would result in removing src contents
      // and therefore a broken symlink would be created.
      stat(src, (err, destStat) => {
        if (err) return cb(err);
        if (destStat.isDirectory() && isSrcSubdir(resolvedDest, resolvedSrc)) {
          return cb(new ERR_FS_COPY_SYMLINK_TO_SUBDIRECTORY({
            code: 'EINVAL',
            message: `cannot overwrite ${resolvedDest} with ${resolvedSrc}`,
            path: dest,
            syscall: 'copy',
            errno: EINVAL,
          }));
        }
        return copyLink(resolvedSrc, dest, cb);
      });
```

And now have been able to exercise the code in test 👍

Reply:
cc @manidlou we should patch this ourselves

Reply:
@bcoe Looking at this code, trying to backport it to fs-extra (https://github.com/jprichardson/node-fs-extra/issues/918); and this doesn't look correct. You're checking if src is a directory: https://github.com/bcoe/node-1/blob/419669cb7946958047c24174d188b8cf06ff7836/lib/internal/fs/cp/cp.js#L374-L375 However, it doesn't matter if src is a directory or not, what you care about is if src is inside dest (assuming dest is a directory). You'd want to check if dest is a directory (via actually `stat`ing dest). On the other hand, you probably don't need to test stats at all, since `isSrcSubdir(dest, src)` shouldn't return `true` unless dest actually is a directory.

Reply:
@RyanZim I haven't had much time in recent months to work on OSS, I'm just seeing this.

It seems like a patch to Node.js would make sense as well for sure. Did you manage to create a regression test?

Reply:
@RyanZim you're correct. The code in node needs to be fixed, and also you're right we don't need to check if dest is a directory as `isSrcSubdir()` would return false in that case anyway.

Reply:
I can't seem to get a regression test working for this; see discussion in https://github.com/jprichardson/node-fs-extra/pull/976.